src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,670 @@
+     *     if the {@code algorithm} or {@code provider} is {@code null}
+     */
+    public static KDF getInstance(String algorithm, Provider provider)
+        throws NoSuchAlgorithmException {
+        Objects.requireNonNull(algorithm, "algorithm must not be null");
+        Objects.requireNonNull(provider, "provider must not be null");
+        try {
+            return getInstance(algorithm, null, provider);
+        } catch (InvalidAlgorithmParameterException e) {
+            throw new NoSuchAlgorithmException(
+                "No implementation found using null KDFParameters", e);
+        }
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm and
+     * is initialized with the specified parameters.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     * @param kdfParameters
+     *     the {@code KDFParameters} used to configure the derivation
+     *     algorithm or {@code null} if no parameters are provided
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDF} implementation for the
+     *     specified algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if no {@code Provider} supports a {@code KDFSpi} implementation for
+     *     the specified algorithm and parameters
+     * @throws NullPointerException
+     *     if the {@code algorithm} is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "algorithm must not be null");
+        // make sure there is at least one service from a signed provider
+        Iterator<Service> t = GetInstance.getServices("KDF", algorithm);
+        NoSuchAlgorithmException lastException = null;
+        while (t.hasNext()) {
+            Service s = t.next();
+            if (!JceSecurity.canUseProvider(s.getProvider())) {
+                continue;
+            }
+            try {
+                Object obj = s.newInstance(kdfParameters);
+                if (!(obj instanceof KDFSpi spiObj)) {
+                    lastException = new NoSuchAlgorithmException(
+                        new InvalidAlgorithmParameterException(
+                            "newInstance failed to provide a KDFSpi for the "
+                            + "provided kdfParameters"));
+                    continue;
+                }
+                if (t.hasNext()) {
```

Comment:
Looks like you can eliminate the `continue;` if you go with if-elses all the way.

```suggestion
                } else if (t.hasNext()) {
```